### PR TITLE
fix: ENTITY_BINDING_SUCCESS event added which is fired whenever there is a successful binding created by the user.

### LINKED ISSUE #20468: [Task]: Track ENTITY_BINDING_SUCCESS
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem
The `BINDING_SUCCESS` event is set to fire every time a widget property is successfully evaluated. This means that the event will be fired on the trigger of actions like "drag and drop", "page refresh", "widget rename", etc... The data from this event is too noisy to derive actual insights. The original purpose of this event was to track successful bindings triggered by user actions on a widget property. 

### Fix
Track a new event named `WIDGET_BINDING_SUCCESS` on successful evaluation based on a config set on the redux store.
Use this map to track the fields which have successful binding

### PR DESCRIPTION
**Pull request by Druthi**
## Description
Adding another event called ENTITY_BINDING_SUCCESS which is fired whenever there is a successful binding created by the user. The BINDING_SUCCESS event was firing more events than actual binding and therefore we created a new event to capture the right data. 

Fixes #20468 

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- Manual

### Issues raised during DP testing
When table widget 'data table' is cleared, ENTITY_BINDING_SUCCESS event is triggered
https://www.loom.com/share/280ab5165b684d59948ae1bc9fe0c074

Templates automatically triggers entitybindingsuccess
https://www.loom.com/share/16be5ae834b44d7bacc73a6d89a99fbd

### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/src/selectors/dataTreeSelectors.ts [hunk 1]
```diff
@@ -66,6 +66,9 @@ export const getUnevaluatedDataTree = createSelector(
 export const getEvaluationInverseDependencyMap = (state: AppState) =>
   state.evaluations.dependencies.inverseDependencyMap;
 
+export const getEvaluationSuccessfulBindindsMap = (state: AppState) =>
```
Comment (ApekshaBhosale) [inline app/client/src/selectors/dataTreeSelectors.ts:69]: do we need to save this in reducers? @Druthi

File: app/client/src/reducers/evaluationReducers/successfulBindingsReducer.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import {
+  ReduxActionTypes,
+  ReduxAction,
+} from "@appsmith/constants/ReduxActionConstants";
+import { createImmerReducer } from "utils/ReducerUtils";
+
+export interface SuccessfulBindingsReduxState {
+  successfulBindings: Record<string, unknown>;
+}
```
Comment (ApekshaBhosale) [inline app/client/src/reducers/evaluationReducers/successfulBindingsReducer.ts:9]: why do we need this?

**Review by ApekshaBhosale — CHANGES_REQUESTED** [review]

File: app/client/src/selectors/dataTreeSelectors.ts (hunk 1 above)
Comment (Druthi) [reply app/client/src/selectors/dataTreeSelectors.ts:69]: How do we save it then to consume for the next evaluation cycle?

File: app/client/src/sagas/PostEvaluationSagas.ts [hunk 3]
```diff
@@ -331,23 +335,60 @@ export function* logSuccessfulBindings(
...
+        }
       }
     }
   });
+
+  if (!successfulBindingsMap) {
+    successfulBindingsMap = new SuccessfulBindingMap(successfulBindingPaths);
+  } else {
+    successfulBindingsMap.setSuccessfulBindings(successfulBindingPaths);
+  }
```
Comment (Rishabh-Rathod) [inline app/client/src/sagas/PostEvaluationSagas.ts:391]: Will there be a use case where we need to generate multiple instances of the class? 
If not in that case, why not create a class with static methods and directly consume `SuccessfulBindingMap.set()` and `SuccessfulBindingMap.get()`?

File: app/client/src/sagas/PostEvaluationSagas.ts [hunk 4]
```diff
@@ -313,11 +318,10 @@ export function* logSuccessfulBindings(
   if (appMode === APP_MODE.PUBLISHED) return;
   if (!evaluationOrder) return;
 
-  if (isCreateFirstTree) {
-    // we only aim to log binding success which were added by user
-    // for first evaluation, bindings are not added by user hence skipping it.
-    return;
-  }
+  const successfulBindingPaths: any = !successfulBindingsMap
```
Comment (Rishabh-Rathod) [inline app/client/src/sagas/PostEvaluationSagas.ts:321]:
```suggestion
  const successfulBindingPaths: unknown = !successfulBindingsMap
```

**Comment by Druthi** [comment]
Updated DP: https://appsmith-q26b6hggx-get-appsmith.vercel.app/applications

File: app/client/src/sagas/PostEvaluationSagas.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ApekshaBhosale) [reply app/client/src/sagas/PostEvaluationSagas.ts:331]: did you add type here?

**Review by ApekshaBhosale — CHANGES_REQUESTED** [review]

**Review by ApekshaBhosale — APPROVED** [review]

**Comment by shwetha-ramesh** [comment]
Working as expected. Moving to Done.

**Review by ApekshaBhosale — APPROVED** [review]

**Comment by KelvinOm** [comment]
Please update the branch from the release, it is necessary for the added checks in the release to start working.

**Pull request merged**